external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/CachedKafkaProducerSuite.scala

```diff
@@ -35,43 +38,68 @@ class CachedKafkaProducerSuite extends SharedSQLContext with PrivateMethodTester
+    val cacheMap = getCacheMap()
+    assert(cacheMap.size === 1)
   }
 
   test("Should close the correct kafka producer for the given kafkaPrams.") {
-    val kafkaParams = new ju.HashMap[String, Object]()
-    kafkaParams.put("acks", "0")
-    kafkaParams.put("bootstrap.servers", "127.0.0.1:9022")
-    kafkaParams.put("key.serializer", classOf[ByteArraySerializer].getName)
-    kafkaParams.put("value.serializer", classOf[ByteArraySerializer].getName)
+    val kafkaParams = getKafkaParams()
     val producer: KP = CachedKafkaProducer.getOrCreate(kafkaParams)
     kafkaParams.put("acks", "1")
     val producer2: KP = CachedKafkaProducer.getOrCreate(kafkaParams)
     // With updated conf, a new producer instance should be created.
     assert(producer != producer2)
-
-    val cacheMap = PrivateMethod[ConcurrentMap[Seq[(String, Object)], KP]]('getAsMap)
-    val map = CachedKafkaProducer.invokePrivate(cacheMap())
-    assert(map.size == 2)
+    val cacheMap = getCacheMap()
+    assert(cacheMap.size === 2)
 
     CachedKafkaProducer.close(kafkaParams)
-    val map2 = CachedKafkaProducer.invokePrivate(cacheMap())
-    assert(map2.size == 1)
-    import scala.collection.JavaConverters._
-    val (seq: Seq[(String, Object)], _producer: KP) = map2.asScala.toArray.apply(0)
-    assert(_producer == producer)
+    val cacheMap2 = getCacheMap()
+    assert(cacheMap2.size === 1)
+    assert(getCacheMapItem(cacheMap2, 0) === producer)
+  }
+
+  test("Should return new instance on calling getOrCreate with same params but task retry.") {
+    val kafkaParams = getKafkaParams()
+    val taskContext = new TaskContextImpl(0, 0, 0, 0, attemptNumber = 0, null, null, null)
+    TaskContext.setTaskContext(taskContext)
+    val producer = CachedKafkaProducer.getOrCreate(kafkaParams)
+    val retryTaskContext = new TaskContextImpl(0, 0, 0, 0, attemptNumber = 1, null, null, null)
+    TaskContext.setTaskContext(retryTaskContext)
+    val producer2 = CachedKafkaProducer.getOrCreate(kafkaParams)
+    assert(producer != producer2)
+    val cacheMap = getCacheMap()
+    assert(cacheMap.size === 1)
+    assert(getCacheMapItem(cacheMap, 0) === producer2)
+  }
+
+  private def getKafkaParams(): ju.HashMap[String, Object] = {
+    val kafkaParams = new ju.HashMap[String, Object]()
+    kafkaParams.put("acks", "0")
+    // Here only host should be resolvable, it does not need a running instance of kafka server.
+    kafkaParams.put("bootstrap.servers", "127.0.0.1:9022")
+    kafkaParams.put("key.serializer", classOf[ByteArraySerializer].getName)
+    kafkaParams.put("value.serializer", classOf[ByteArraySerializer].getName)
+    kafkaParams
+  }
+
+  private def getCacheMap(): ConcurrentMap[Seq[(String, Object)], KP] = {
+    val getAsMap = PrivateMethod[ConcurrentMap[Seq[(String, Object)], KP]]('getAsMap)
+    CachedKafkaProducer.invokePrivate(getAsMap())
+  }
+
+  private def getCacheMapItem(map: ConcurrentMap[Seq[(String, Object)], KP], offset: Int): KP = {
```

Comment:
Hmm... maps don't necessarily have deterministic iteration order, so this method only really makes sense if the map has a single item. Since you always call it with `0` as the offset anyway, it'd be better to just simplify it (`map.values().iterator().next()`).

Or maybe explicitly using `map.get(kafkaParams)` in the tests.